### PR TITLE
json schema layer

### PR DESCRIPTION
**Pull request by kingsleyzissou**
##### Checklist

- [ ] `npm test` passes
- [ ] `npm run build` works
- [ ] tests are included
- [ ] documentation is changed or added

### CODE REVIEW

**Comment by kingsleyzissou** [comment]
This is still the early stages. The next step is to bring this into the datastore and create this layer before passing this to the model. I need to sort out some of the typings before I do that.

TODO:
- write some tests for this
- complete validation method or otherwise keep validation per field level as it currently is
- add the comments for the types and the methods
- hook this up to the rest of the datastore

I am just creating the PR for a bit of visibility and early feedback

@Eunovo @wtrocki

**Comment by wtrocki** [comment]
Perfecto. This PR has to go first so I will deal with possible conflicts on my side

File: packages/offix-datastore/src/ModelSchema.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
+import invariant from "tiny-invariant";
+import { JSONSchema7 } from "json-schema";
+
+export interface DSProperties extends JSONSchema7 {
```
Comment (wtrocki) [inline packages/offix-datastore/src/ModelSchema.ts:4]: Not particularly like those names

File: packages/offix-datastore/src/ModelSchema.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  index?: boolean;
+  primary?: boolean;
+  default?: any;
+  encrypted?: boolean;
+  key?: string;
```
Comment (wtrocki) [inline packages/offix-datastore/src/ModelSchema.ts:9]: What is key?

File: packages/offix-datastore/src/ModelSchema.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+export declare class DSJsonSchema<T> {
+  name: string;
+  namespace?: string;
+  version: number;
+  indices?: string[];
```
Comment (wtrocki) [inline packages/offix-datastore/src/ModelSchema.ts:16]: Indexes sound better?

File: packages/offix-datastore/src/ModelSchema.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+  encrypted?: string[];
+  primaryKey?: string;
+  type: "object";
```
Comment (wtrocki) [inline packages/offix-datastore/src/ModelSchema.ts:19]: Create type and add scalar as well

Comment (kingsleyzissou) [reply packages/offix-datastore/src/ModelSchema.ts:19]: @wtrocki type is there already, will add scalar

**Pull request merged**